### PR TITLE
Add check for second value in sum: Logsumexp

### PR DESCRIPTION
**Pull request by shivam096**
Added conditions to check if second value exists in "sum" for Logsumexp update. If not then no change.

Files updated:

- torchfix/visitors/misc/__init__.py
- tests/fixtures/misc/checker/logsumexp.py

### CODE REVIEW

File: torchfix/visitors/misc/__init__.py [hunk 1]
```diff
@@ -183,7 +183,7 @@ def visit_Call(self, node):
                             node.args[0].value.args[0].value
                         )
                         == "torch.exp"
-                    ):
+                    ) and len(node.args[0].value.args) > 1 and node.args[0].value.args[1].value is not None:
```
Comment (kit1980) [inline torchfix/visitors/misc/__init__.py:186]: Use `get_specific_arg` from https://github.com/pytorch-labs/torchfix/blob/main/torchfix/common.py#L64 because `dim` can be passed as a positional or as a keyword argument.

File: tests/fixtures/misc/checker/logsumexp.py [hunk 2]
```diff
@@ -12,3 +12,4 @@
 y = torch.log(2 + x)
 y = torch.sum(torch.log(torch.exp(x)), 1)
 y = torch.exp(torch.sum(torch.log(x), 1, keepdim=True))
+y = torch.log(torch.sum(torch.exp(2.5)))
```
Comment (kit1980) [inline tests/fixtures/misc/checker/logsumexp.py:15]: Add a comment explaining why it should not be triggered.

Comment (kit1980) [inline tests/fixtures/misc/checker/logsumexp.py:15]: add

`y = torch.sum(torch.log(torch.exp(x)), dim=1)` - should trigger
`y = torch.sum(torch.log(torch.exp(x)), None)` - should not trigger
`y = torch.sum(torch.log(torch.exp(x)), dim=None)` - should not trigger

File: torchfix/visitors/misc/__init__.py (hunk 1 above)
Comment (shivam096) [reply torchfix/visitors/misc/__init__.py]: Updated and added condition to check for dim. But I used has_specific_arg since we just want to check its presence

File: tests/fixtures/misc/checker/logsumexp.py (hunk 2 above)
Comment (shivam096) [reply tests/fixtures/misc/checker/logsumexp.py]: Added comment

Comment (shivam096) [reply tests/fixtures/misc/checker/logsumexp.py]: Updated the test cases

File: torchfix/visitors/misc/__init__.py [hunk 3]
```diff
@@ -184,9 +184,13 @@ def visit_Call(self, node):
                         )
                         == "torch.exp"
                     ):
-                        self.add_violation(
-                            node,
-                            error_code=self.ERRORS[0].error_code,
-                            message=self.ERRORS[0].message(),
-                            replacement=None,
-                        )
+                        if len(node.args[0].value.args) > 1 and (
```
Comment (kit1980) [inline torchfix/visitors/misc/__init__.py:187]: I don't think `value.args[1]` is always correct, because the order may be different for keyword arguments.
For example, add test  `y = torch.sum(torch.log(torch.exp(x)), keepdim=True, dim=None)`.

I think you can delete most if this and just use `get_specific_arg("dim", 1)`, which will return None if the argument is not present.

File: tests/fixtures/misc/checker/logsumexp.py [hunk 4]
```diff
@@ -5,10 +5,16 @@
 # logsumexp
 y = torch.log(torch.sum(torch.exp(x), 1, keepdim=True))
 y = torch.log(torch.sum(torch.exp(2.5 + x), 1))
+y = torch.log(torch.sum(torch.exp(1),1,1))
+y = torch.log(torch.sum(torch.exp(1),1,dim=1))
+y = torch.log(torch.sum(torch.exp(x), keepdim=True, dim=None))
```
Comment (kit1980) [inline tests/fixtures/misc/checker/logsumexp.py:10]: This should not trigger.
`dim=None` is the same as no `dim` at all.

File: tests/fixtures/misc/checker/logsumexp.py [hunk 5]
```diff
@@ -5,10 +5,16 @@
...
 
 # not logsumexp
 y = torch.log(torch.sum(torch.exp(x), 1, keepdim=True) + 2.5)
 y = torch.log(torch.sum(torch.exp(x) + 2.5, 1))
 y = torch.log(2 + x)
 y = torch.sum(torch.log(torch.exp(x)), 1)
 y = torch.exp(torch.sum(torch.log(x), 1, keepdim=True))
+y = torch.log(torch.sum(torch.exp(2.5))) # this should not be flagged as the second argument is missing for sum function call
+y = torch.sum(torch.log(torch.exp(x)), dim=1) 
+y = torch.sum(torch.log(torch.exp(x)), dim=None)
```
Comment (kit1980) [inline tests/fixtures/misc/checker/logsumexp.py:20]: Change to the order of calls to `log(sum(exp()))` as we discussed.

File: torchfix/visitors/misc/__init__.py [hunk 6]
```diff
@@ -184,9 +184,12 @@ def visit_Call(self, node):
                         )
                         == "torch.exp"
                     ):
-                        self.add_violation(
-                            node,
-                            error_code=self.ERRORS[0].error_code,
-                            message=self.ERRORS[0].message(),
-                            replacement=None,
-                        )
+                        if self.get_specific_arg(
```
Comment (kit1980) [inline torchfix/visitors/misc/__init__.py:187]: Even if the `dim` arg is present, you need to check that its value is not `None`.

File: tests/fixtures/misc/checker/logsumexp.py (hunk 5 above)
Comment (shivam096) [reply tests/fixtures/misc/checker/logsumexp.py:20]: I thought these were false test cases. Will update it

File: torchfix/visitors/misc/__init__.py (hunk 3 above)
Comment (shivam096) [reply torchfix/visitors/misc/__init__.py]: Updated the code and resolved

File: tests/fixtures/misc/checker/logsumexp.py (hunk 4 above)
Comment (shivam096) [reply tests/fixtures/misc/checker/logsumexp.py]: Updated the condition to resolve it

File: torchfix/visitors/misc/__init__.py (hunk 6 above)
Comment (shivam096) [reply torchfix/visitors/misc/__init__.py:187]: Added condition to check for None value. Since the value type is string hence performed a string comparision

File: torchfix/visitors/misc/__init__.py [hunk 7]
```diff
@@ -184,9 +184,18 @@ def visit_Call(self, node):
...
+                            node.args[0].value, arg_name="dim", arg_pos=1
+                        ):
+                            if (
+                                self.get_specific_arg(
+                                    node.args[0].value, arg_name="dim", arg_pos=1
+                                ).value.value
```
Comment (kit1980) [inline torchfix/visitors/misc/__init__.py:193]: You only check for value of the argument when it's present.
It should be two nested if's - first if it's present, then if value is not None.

File: tests/fixtures/misc/checker/logsumexp.py [hunk 8]
```diff
@@ -3,12 +3,18 @@
...
-y = torch.exp(torch.sum(torch.log(x), 1, keepdim=True))
+y = torch.log(torch.sum(torch.exp(x), 1, keepdim=True) + 2.5) # cant have an addition operation inside the log function call 
+y = torch.log(torch.sum(torch.exp(x) + 2.5, 1)) # Cant have an addition operation inside the sum function call with the argument as it expects a tensor
+y = torch.log(2 + x) # missing sum and exp
+y = torch.sum(torch.log(torch.exp(x)), 1) # not proper order of log and sum
+y = torch.exp(torch.sum(torch.log(x), 1, keepdim=True)) #order of log,sum and exp is reversed
+y = torch.log(torch.sum(torch.exp(2.5))) # this should not be flagged as the second argument is missing for sum function call and exp function call has an integer argument instead of a tensor
+y = torch.log(torch.sum(torch.exp(x)), dim=1) #dim is not part of the sum fuction call
+y = torch.log(torch.sum(torch.exp(x)), dim=None) #dim is not part of the sum fuction call and dim is None
+y = torch.log(torch.sum(torch.exp(x), keepdim=True, dim=None)) #dim argument cannot be None
```
Comment (kit1980) [inline tests/fixtures/misc/checker/logsumexp.py:20]: Neet a test when `dim` is not provided. All these have `dim` provided currently, thus missing a bug in the code.

File: torchfix/visitors/misc/__init__.py (hunk 7 above)
Comment (shivam096) [reply torchfix/visitors/misc/__init__.py:193]: The first "if" condition on line 187 checks for the presence of "dim" and then if confirmed it is moved to the second "if" in line 190.

File: tests/fixtures/misc/checker/logsumexp.py (hunk 8 above)
Comment (kit1980) [reply tests/fixtures/misc/checker/logsumexp.py:20]: Nevermind, I see there is a test.

File: torchfix/visitors/misc/__init__.py (hunk 7 above)
Comment (kit1980) [reply torchfix/visitors/misc/__init__.py:193]: I see now.
You should assign the arg to a variable to reduce code duplication.
And then add `assert` that it's not `None`, otherwise MyPy is complaining:

https://github.com/pytorch-labs/torchfix/actions/runs/13081681874/job/36506448560?pr=90

See this for example https://github.com/pytorch-labs/torchfix/blob/main/torchfix/visitors/deprecated_symbols/__init__.py#L35

File: tests/fixtures/misc/checker/logsumexp.py (hunk 8 above)
Comment (shivam096) [reply tests/fixtures/misc/checker/logsumexp.py:20]: Is my assumption correct that the following transformation should not be flagged -

`y = torch.log(torch.sum(torch.exp(x)))`

File: torchfix/visitors/misc/__init__.py (hunk 7 above)
Comment (kit1980) [reply torchfix/visitors/misc/__init__.py:193]: You may also need to use `ensure_type`, like here: https://github.com/pytorch-labs/torchfix/blob/main/torchfix/visitors/deprecated_symbols/qr.py#L19

Please run mypy locally and verify it passes.

File: tests/fixtures/misc/checker/logsumexp.py (hunk 8 above)
Comment (kit1980) [reply tests/fixtures/misc/checker/logsumexp.py:20]: @shivam096 yes, as `dim` is not present.

File: torchfix/visitors/misc/__init__.py (hunk 7 above)
Comment (shivam096) [reply torchfix/visitors/misc/__init__.py:193]: I ran the code and mypy errors are only resolved when I do a `isInstance()` check. 

And in doing so I need to check for both `Integer` and `Tuple` since the `dim` can have both of the types as its argument value and not of type `Name` which is there when `None` value is there.

And since value of tuple cannot be retrieved through `.value`. I need to update the code to different code structure that could handle both integer and tuple values for `dim` which are Integers and are not `None`.

Comment (shivam096) [reply torchfix/visitors/misc/__init__.py]: I have updated the code and have made the necessary changes to handle any future type based issues. 

Updated the test cases as well

File: torchfix/visitors/misc/__init__.py [hunk 9]
```diff
@@ -184,9 +184,46 @@ def visit_Call(self, node):
                         )
                         == "torch.exp"
                     ):
-                        self.add_violation(
-                            node,
-                            error_code=self.ERRORS[0].error_code,
-                            message=self.ERRORS[0].message(),
-                            replacement=None,
-                        )
+                        if (
```
Comment (kit1980) [inline torchfix/visitors/misc/__init__.py:187]: This is already checked right in the previous lines, why duplicate?

File: torchfix/visitors/misc/__init__.py [hunk 10]
```diff
@@ -184,9 +184,46 @@ def visit_Call(self, node):
...
+                                self.get_qualified_name_for_call(
+                                    node.args[0].value.args[0].value
+                                )
+                                == "torch.exp"
+                            ):
+                                dim_arg = self.get_specific_arg(
+                                    node.args[0].value, arg_name="dim", arg_pos=1
+                                )
+                                if dim_arg:  # checks if dim argument is present
+                                    if isinstance(
```
Comment (kit1980) [inline torchfix/visitors/misc/__init__.py:201]: This lines are redundant, no?
Later there are checks for `cst.Integer` and `cst.Tuple`.

File: torchfix/visitors/misc/__init__.py [hunk 11]
```diff
@@ -184,9 +184,46 @@ def visit_Call(self, node):
...
+                                        dim_arg.value, cst.Integer
+                                    ) or isinstance(
+                                        dim_arg.value, cst.Tuple
+                                    ):  # checks if dim argument is an integer or tuple
+                                        if (
+                                            isinstance(dim_arg.value, cst.Integer)
```
Comment (kit1980) [inline torchfix/visitors/misc/__init__.py:207]: `cst.Integer` can not be None, meaningless condition.

File: torchfix/visitors/misc/__init__.py [hunk 12]
```diff
@@ -184,9 +184,46 @@ def visit_Call(self, node):
...
+                                            and dim_arg.value.value != "None"
+                                        ):
+                                            self.add_violation(
+                                                node,
+                                                error_code=self.ERRORS[0].error_code,
+                                                message=self.ERRORS[0].message(),
+                                                replacement=None,
+                                            )
+                                        elif isinstance(
+                                            dim_arg.value, cst.Tuple
```
Comment (kit1980) [inline torchfix/visitors/misc/__init__.py:217]: How can it be `cst.Tuple`?

File: tests/fixtures/misc/checker/logsumexp.py [hunk 13]
```diff
@@ -3,12 +3,21 @@
...
+y = torch.log(torch.sum(torch.exp(x) + 2.5, 1)) # Cant have an addition operation inside the sum function call with the argument as it expects a tensor
+y = torch.log(2 + x) # missing sum and exp
+y = torch.sum(torch.log(torch.exp(x)), 1) # not proper order of log and sum
+y = torch.exp(torch.sum(torch.log(x), 1, keepdim=True)) #order of log,sum and exp is reversed
+y = torch.log(torch.sum(torch.exp(2.5))) # this should not be flagged as the second argument is missing for sum function call and exp function call has an integer argument instead of a tensor
+y = torch.log(torch.sum(torch.exp(x)), dim=1) #dim is not part of the sum fuction call
+y = torch.log(torch.sum(torch.exp(x)), dim=None) #dim is not part of the sum fuction call and dim is None
+y = torch.log(torch.sum(torch.exp(x), keepdim=True, dim=None)) #dim argument cannot be None 
+y = torch.log(torch.sum(torch.exp(x), dim=(1,None))) #dim argument cannot be a tuple with None
+y = torch.log(torch.sum(torch.exp(x), dim=(None,None))) #dim argument cannot be a tuple with None
```
Comment (kit1980) [inline tests/fixtures/misc/checker/logsumexp.py:23]: No need to check for `dim=(None,None)` or `dim=(1,None)`, it can not happen because if present it's an int or tuple of ints: https://pytorch.org/docs/stable/generated/torch.sum.html

**Comment by kit1980** [comment]
https://github.com/pytorch-labs/torchfix/pull/91

File: torchfix/visitors/misc/__init__.py (hunk 9 above)
Comment (shivam096) [reply torchfix/visitors/misc/__init__.py:187]: Error code while testing. Removed it

File: torchfix/visitors/misc/__init__.py (hunk 10 above)
Comment (shivam096) [reply torchfix/visitors/misc/__init__.py:201]: Yes. Removed it since they were test code.

File: torchfix/visitors/misc/__init__.py (hunk 11 above)
Comment (shivam096) [reply torchfix/visitors/misc/__init__.py:207]: Here the condition checks if the value is of type integer and also makes sure that the value it holds is also not None since Tuples in `dim` cannot have None values

**Comment by kit1980** [comment]
closing this in favor of https://github.com/pytorch-labs/torchfix/pull/91
